feat: dependency set refactor (#19)
defi-wonderland/specs: specs/protocol/superchain-configuration.md

- An immutable address pointing to the `SharedLockbox` contract.
- This address MUST be immutable because there's only one `SharedLockbox` for each cluster.

#### `dependencyManager`

- An address with the ability to add a dependency to the dependency set.
- The `dependencyManager` can only be set during initialization.

#### `dependencySet`

- Stores the current list of chain IDs in the op-governed dependency set.
- It MUST contain all the chain IDs of the chains that integrate the corresponding Superchain network.

#### `addDependency`

The `addDependency` function adds a new chain to the op-governed cluster.

It can only be called by the `DEPENDENCY_MANAGER` role in the `SuperchainConfig` and ensures that the chain ID
is not already included in the dependency set.

Before proceeding, it verifies that the dependency set is not full (`type(uint8).max`).

> The function allowlists the new chain's `OptimismPortal` in the `SharedLockbox`.

Reviewer: ```suggestion
This function allowlists the new chain's `OptimismPortal` in the `SharedLockbox`.
```

Finally, it emits the `DependencyAdded` event with the `chainId`, its corresponding `SystemConfig`, and `OptimismPortal`.

```solidity
function addDependency(uint256 _chainId, address _systemConfig) external;
```

### Events

#### `DependencyAdded`

MUST be triggered when `addDependency` is called

```solidity
event DependencyAdded(uint256 indexed chainId, address indexed systemConfig, address indexed portal);
```

### Invariants

- Only the `DEPENDENCY_MANAGER` role MUST be able to add a new dependency to the dependency set

> - The dependency set MUST NOT be at it's max size before adding a new dependency

Reviewer: ```suggestion
- A new chain cannot be added to the dependency set once the maximum size is reached.
```

> - The chain ID MUST be different to the current `block.chainid`

Reviewer: ```suggestion
- The chain ID MUST be different from the current `block.chainid`
```
